File: leetcode.com/0279_perfect_squares/solution.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

#define MIN(a, b)   (a < b) ? a : b
#define MAX_LENGTH  10000

int minimum[MAX_LENGTH];

int numSquares(int n)
{
    /* pre compute values */
    if (minimum[1] == 0) {
        for (int i = 1 ; i < MAX_LENGTH ; i++) {
            int ps = (int) floor(sqrt(i));
            int delta = i - ps * ps;

            if (delta == 0) {
                minimum[i] = 1;
            } else {
                minimum[i] = minimum[ps * ps] + minimum[delta];

                for (int j = ps ; j > 0 ; j--) {
                    int delta = i - j * j;

                    minimum[i] = MIN(
                        minimum[i],
                        minimum[j * j] + minimum[delta]
                    );
                }
            }
        }
    }

    return minimum[n];
}

int main()
{
    printf("%d == 4\n", numSquares(7));
    return 0;
}

File: leetcode.com/0279_perfect_squares/solution.go
package main

import "fmt"
import "math"

var values [10000]int

func min(a, b int) int {
    if (a < b) {
        return a
    }

    return b
}

func numSquares(n int) int {
    if values[1] != 0 {
        return values[n]
    }

    for v := 1 ; v < len(values) ; v++ {
        ps := int(math.Floor(math.Sqrt(float64(v))))
        delta := v - ps*ps

        if delta == 0 {
            values[v] = 1
        } else {
            values[v] = values[ps*ps] + values[delta]

            for s := ps - 1 ; s > 0 ; s-- {
                delta := v - s*s
                values[v] = min(values[v], values[s*s] + values[delta])
            }
        }
    }

    return values[n]
}

func main() {
    fmt.Println(numSquares(12))
}
